fabtests/functional/rdm_tagged_peek.c

```diff
@@ -89,12 +71,11 @@ static int trecv_op(uint64_t tag, uint64_t flags, bool ignore_nomsg)
 			FT_PRINTERR("fi_trecvmsg", ret);
 			return ret;
 		}
-
-		ret = fi_cq_sread(rxcq, &comp, 1, NULL, -1);
-		if (ret == 1) {
-			ret = 0;
+		ft_tag = tag;
+		rx_cq_cntr = 0;
+		ret = ft_read_cq(rxcq, &rx_cq_cntr, 1, timeout);
+		if (ret == FI_SUCCESS)
```

Comment:
Can we change wait_for_send_comp() to wait_for_comp(), pass in the cq to check, and call it from here?

Reply:
So you want a single `wait_for_comp()` that can handle both send/recv completion in this test, right?

Reply:
The change below to call ft_get_tx_comp() directly instead of wait_for_send_comp() makes sense.  Maybe we also need ft_get_rx_comp() as a peer?  It's the direct setting of ft_tag and rx_cq_cntr that look odd here.  (And something that we someday need to address to make these tests easier to maintain.

Reply:
I thought about using `ft_get_rx_comp`, one thing a bit tricky is `trecv_op` could retry the trecv and cq poll when `ignore_nomsg` is true, which will makes the rx_seq/rx_cq_cntr messed up. About the ft_tag set,  AFAICT it is set as 0 by default in `shared.c` and overridden by test code during the test. The common code in `shared.c` will use it  when it is non-zero otherwise it will use tx/rx_cq_cntr as the tag. I am not sure whether it is the expected workflow for using `ft_tag` but it really seems foggy to me. Should we make tag an argument of `ft_post_tx/rx_buf` instead. `ft_tx`/`ft_rx` can still use cq cntr as tags, but it leaves callers flexibility to customize the tag.

Reply:
As for a single wait_for_comp() -- don't add that yet (and maybe not)

Reply:
Also `ft_get_rx_comp` will always log error if there is any. But the test wants to ignore enomesg error if requested. That is why I break `ft_get_*_comp` to two parts, 1 for `ft_read_cq` and the other for `fi_cq_readerr`. If I use `ft_get_rx_comp` in recv path I will lose this flexibility, do you have better ideas here?

Reply:
The global tag is likely a mistake.  I haven't looked into details as to how to start restructuring fabtests, but I know the current approach makes it difficult to isolate test changes.  If you think it helps adding tag as a parameter to some of the lower-level calls and using those directly, I'd say go for it.  We can review those changes until we're happy with them, then look at adopting them to other tests.

Reply:
I don't have better ideas.  I don't think the problems are in your changes or this test, but the common code.  And there's no easy way to handle that without a deep dive into serious restructuring of the code.  I think we accept the above changes for now and figure out how to handle a broader cleanup later.

Reply:
> If you think it helps adding tag as a parameter to some of the lower-level calls and using those directly

Thanks. I think I will try in this direction to get rid of `ft_tag` and the shenanigans with the tx/rx_seq, cntr at the cleanup stage. I agree that the current implementation is hard for maintenance.

Reply:
@shefty  I think I had a simpler solution in the latest revision so I can get rid all the hack on `ft_tag`, `tx/rx_seq` and `tx/rx_cq_cntr`. Please see the latest revision.